hkmc2/shared/src/main/scala/hkmc2/semantics/Term.scala

```diff
@@ -44,79 +50,83 @@ sealed trait ResolvableImpl:
    */
   private var expansion: Opt[Opt[Term]] = N
 
-  def duplicate: Resolvable =
-    t.match
+  def duplicate: this.type =
+    this.match
       case t: Term.Ref => t.copy()(t.tree, t.refNum, t.resSym)
       case t: Term.App => t.copy()(t.tree, t.sym, t.resSym)
       case t: Term.TyApp => t.copy()(t.sym)
       case t: Term.Sel => t.copy()(t.sym)
       case t: Term.SynthSel => t.copy()(t.sym)
-    .withLocOf(t)
+    .withLocOf(this)
+    .asInstanceOf
   
   override def show: Str = expansion match
-    case S(S(expansion)) => t.showDbg + "{~>" + expansion.show + "}"
-    case _ => t.showDbg
+    case S(S(expansion)) => showDbg + "{~>" + expansion.show + "}"
+    case _ => showDbg
+  
+  // Scala requires `abstract override` for this to work. Only god knows why.
+  abstract override def describe: Str = 
+    val desc = super.describe // this calls Term.describe
+    defn match
+    case S(df: TermDefinition) =>
+      s"${desc} with a ${df.k.desc} definition '${df.sym.nme}'"
+    case S(df: ClassLikeDef) =>
+      s"${desc} with a ${df.kind.desc} definition '${df.sym.nme}'"
+    case S(df: TypeDef) =>
+      s"${desc} with a type definition '${df.sym.nme}'"
+    case N =>
+      s"${desc} without a resolvable definition"
   
   def instantiate = expansion match
     case S(S(t)) => t
-    case S(N) => t
-    case N => lastWords(s"missing expansion for term ${t}")
+    case S(N) => this
+    // case N => lastWords(s"missing expansion for term ${this}")
+    case N =>
+      // FIXME: instantiating a term without an expansion should be an
+      // error
+      this
 
   /** This method is only supposed to be called by Resolver. */
-  private[semantics] def expand(expansionFn: Opt[Term => Term]): this.type =
-    val newExpansion = expansionFn.map(_(t.duplicate.resolve))
+  private[semantics] def expand(expansionFn: Opt[this.type => Term]): this.type =
```

Comment:
Why does `expand` take a function?

Reply:
That's a safety measure that I initially took to prevent from forgetting `duplicate`ing the term. Now it seems that it doesn't add much safety... I think we should remove it and accept `Term` instead.